python/test/unit/language/test_pipeliner.py

```diff
@@ -33,32 +33,51 @@ def check_capabilities():
 
 @triton.jit
 def matmul_kernel(  #
-        a_ptr, b_ptr, output_ptr,  #
+        a_ptr, scale_ptr, b_ptr, output_ptr,  #
         M, N, K,  #
         stride_am, stride_ak,  #
+        stride_sm, stride_sk,  #
         stride_bk, stride_bn,  #
         stride_cm, stride_cn,  #
         BLOCK_M: tl.constexpr, BLOCK_N: tl.constexpr, BLOCK_K: tl.constexpr,  #
-        NUM_STAGES: tl.constexpr):
+        NUM_STAGES: tl.constexpr, a_type: tl.constexpr, b_type: tl.constexpr):
     pid = tl.program_id(axis=0)
     num_pid_m = tl.cdiv(M, BLOCK_M)
     pid_m = pid % num_pid_m
     pid_n = pid // num_pid_m
     offs_am = (pid_m * BLOCK_M + tl.arange(0, BLOCK_M)) % M
     offs_bn = (pid_n * BLOCK_N + tl.arange(0, BLOCK_N)) % N
+    IS_SCALED: tl.constexpr = a_type is not None and b_type is not None
+    DIV_FACTOR: tl.constexpr = 2 if IS_SCALED and a_type == "e2m1" else 1
+    KA: tl.constexpr = BLOCK_K // DIV_FACTOR
+    BLOCK_AK: tl.constexpr = BLOCK_K // DIV_FACTOR
     offs_k = tl.arange(0, BLOCK_K)
-    a_ptrs = a_ptr + (offs_am[:, None] * stride_am + offs_k[None, :] * stride_ak)
+    offs_ak = tl.arange(0, BLOCK_AK)
+    a_ptrs = a_ptr + (offs_am[:, None] * stride_am + offs_ak[None, :] * stride_ak)
     b_ptrs = b_ptr + (offs_k[:, None] * stride_bk + offs_bn[None, :] * stride_bn)
+    if IS_SCALED:
+        BLOCK_SK: tl.constexpr = BLOCK_K // 32
+        offs_sk = tl.arange(0, BLOCK_SK)
+        scale_ptrs = scale_ptr + (offs_am[:, None] * stride_sm + offs_sk[None, :] * stride_sk)
     accumulator = tl.zeros((BLOCK_M, BLOCK_N), dtype=tl.float32)
     for k in tl.range(0, tl.cdiv(K, BLOCK_K), num_stages=NUM_STAGES):
-        mask_a = (offs_am[:, None] < M) & (offs_k[None, :] + k * BLOCK_K < K)
+        mask_a = (offs_am[:, None] < M) & (offs_ak[None, :] + k * BLOCK_AK < KA)
         mask_b = ((offs_k[:, None] + k * BLOCK_K) < K) & (offs_bn[None, :] < N)
         a = tl.load(a_ptrs, mask=mask_a, other=0)
         b = tl.load(b_ptrs, mask=mask_b, other=0)
-        accumulator = tl.dot(a, b, acc=accumulator)
-        a_ptrs += BLOCK_K * stride_ak
+        if IS_SCALED:
+            # Adapted scale indexing and dot_scaled operation
+            mask_scale = (offs_am[:, None] < M) & (offs_sk[None, :] + k * BLOCK_SK < K // 32)
+            a_scale = tl.load(scale_ptrs, mask=mask_scale, other=0)
```

Comment:
This load is not pipelined at the moment as it's "too small" (loading just 2 elements in the fast-running access), and perhaps this is alright? @ThomasRaoux thoughts? Otherwise I can special-case.

The better fix would be to be able to vectorise loads from shared memory across dimensions, but we don't currently implement that.

Reply:
right, I think based on performance we can special case it for the upscale case or maybe it is good enough. 

> The better fix would be to be able to vectorise loads from shared memory across dimensions, but we don't currently implement that.

Yes that would be good but I'm not sure that would cause us to pipeline this?

Reply:
as per internal discussion, we'll figure out whether to pipeline this one or not once we start benchmarking. We should also make sure that, even if we don't pipeline it, we don't have a spurious layout conversion.